Katana allocator variants for Sets and Maps

l-hoang opened the pull request:
Thsi commit fixes the allocator for gstl's Unordered Map and adds a gstl
variant for unordered set.
This is needed for some graph neural network related code coming down the line.

Also changes allocator for Set/Map to Pow2Alloc as well.

ddn0 commented on .github/workflows/conda.yaml:
```diff
@@ -40,12 +40,11 @@ jobs:
-        pushd external/libcypher-parser
-        git apply ../../scripts/cypher-parser.patch
-        popd
+        # TODO calling setup_ubuntu too heavy-weight for this build
```
@arthurp do you have thoughts on how to configure the conda builds?

I would think for conda, this would be the place to install build environment stuff, but actual build dependencies would have to go in the corresponding conda meta.yaml files.

l-hoang replied:
You are correct (fixed the conda build locally and figured out this was how things worked).

arthurp replied:
Are we shipping a patched cypher parser with Galois now? Is cypher parser statically or dynamically linked?

Ideally we would use the [conda-forge libcypher-parser package](https://anaconda.org/rapidsai/libcypher-parser). If we cannot use that then conda packaging will be a bit tricky.

l-hoang replied:
There's currently some discussion going on in slack re: this; there are some build issues we're looking into.

ddn0 commented on cmake/Modules/FindCypherParse.cmake:
```diff
@@ -0,0 +1,18 @@
+
+# set the plural versions of the libraries
+set(CypherParse_INCLUDE_DIRS ${CypherParse_INCLUDE_DIR})
+set(CypherParse_LIBRARIES ${CypherParse_LIBRARY})
```
I'd prefer to use target based builds where users only have to do `target_link_libraries(my-app libcypher)`. 

Can you create a target and use `target_include_directories` and `target_link_libraries` here so that users only have to do something like `target_link_libraries(query cypher)`?

l-hoang commented:
Another current issue is that the Python build fails because I made libcypher-parser a required dependency in CMake, so that fails on the Python packaging test because that test doesn't install many things.

Should I at least make it install libcypher?

roshandathathri commented on libquery/include/querying/CypherCompiler.h:
```diff
@@ -7,12 +7,10 @@
 #include <stack>
 #include "querying/Matching.h"
-// cypher parser includes
+// 3rd party cypher library
```
No need for this comment.

roshandathathri commented on CMakeLists.txt:
```diff
@@ -201,6 +201,9 @@ include(HandleSanitizer)
 include(CheckEndian)
 
+# cypher parser library
+find_package(CypherParse REQUIRED)
```
I would prefer CypherParser or LibCypherParser

l-hoang replied:
I'll go with CypherParser.

roshandathathri commented on lonestar/querying/cpu/querytest/README.md:
```diff
@@ -1,17 +1,11 @@
-`git submodule init; git submodule update`
-to get the Cypher parser dependency. Then,
-`cd external/libcypher-parser; git apply ../../scripts/cypher-parser.patch` to
-fix a build problem with the parser and recent compilers.
```
Add a note about library dependencies

l-hoang replied:
Probably not required because CMake will fail if it doesn't find the library installed.

roshandathathri replied:
Yes, but we need to mention the dependences in our README.

l-hoang replied:
That's already done at the top level README.

l-hoang commented:
Currently facing an issue with the conda build as well: CMake is able to find the include/library for the cypher-parser, but during the actual build, it fails as it is suddenly unable to find cypher-parser.h (found in /usr/include).

Other comments from reviewers have been resolved. Python build is also still an issue.

l-hoang commented:
Everything passes except the python build (which we're dropping anyways in another PR). Thanks for the help everyone! Merging now.

tylershunt reviewed (approved):
Seems fine to me.

roshandathathri commented on libgalois/include/katana/gstl.h:
```diff
@@ -61,14 +62,19 @@ using List = std::list<T, FixedSizeAlloc<T>>;
-using UnorderedMap = std::unordered_map<
-    K, V, Hash, KeyEqual, FixedSizeAlloc<std::pair<const K, V>>>;
+using UnorderedMap =
+    std::unordered_map<K, V, Hash, KeyEqual, Pow2Alloc<std::pair<const K, V>>>;
```
Fix `Map` and `Set` as well to use `Pow2Alloc`.

l-hoang replied:
I'm actually not sure if that works or not from what I remember; unordered vs ordered
containers seem to have different allocator requirements.

I'll test it out and see what happens I suppose.

roshandathathri replied:
Thanks. If it doesn't work, please add a comment saying so.

l-hoang replied:
Seems to work from the bit of testing I did. New commit has it changed.

roshandathathri reviewed (approved):
LGTM

Commit:
Katana allocator for Sets and Maps

Thsi commit fixes the allocator for gstl's Unordered Map and adds a gstl
variant for unordered set.
This is needed for some graph neural network coming down the line.

Also changes Map and Set (ordered variants) to use Pow2Alloc as well.
